File: instabiz/instabiz/report/ib_collections_report/ib_collections_report.py
"""IB Collections Report — per-rep invoice, collected, outstanding, collection %."""
import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
	filters = filters or {}
	columns = _columns()
	data    = _data(filters)
	chart   = _chart(data, filters)
	summary = _summary(data)
	return columns, data, None, chart, summary


def _columns():
	return [
		{"fieldname": "sales_person",     "label": _("Sales Person"),       "fieldtype": "Data",     "width": 180},
		{"fieldname": "invoice_count",    "label": _("Invoices"),            "fieldtype": "Int",      "width": 90},
		{"fieldname": "invoiced_amount",  "label": _("Invoiced (₹)"),        "fieldtype": "Currency", "width": 140},
		{"fieldname": "collected_amount", "label": _("Collected (₹)"),       "fieldtype": "Currency", "width": 140},
		{"fieldname": "outstanding",      "label": _("Outstanding (₹)"),     "fieldtype": "Currency", "width": 140},
		{"fieldname": "collection_pct",   "label": _("Collection %"),        "fieldtype": "Percent",  "width": 110},
	]


def _data(filters):
	conds = ["si.docstatus = 1", "si.is_return = 0", "si.custom_sales_person_user IS NOT NULL"]
	params = {}

	if filters.get("from_date"):
		conds.append("si.posting_date >= %(from_date)s")
		params["from_date"] = filters["from_date"]
	if filters.get("to_date"):
		conds.append("si.posting_date <= %(to_date)s")
		params["to_date"] = filters["to_date"]
	if filters.get("territory"):
		conds.append("si.territory = %(territory)s")
		params["territory"] = filters["territory"]
	if filters.get("sales_person_user"):
		conds.append("si.custom_sales_person_user = %(sales_person_user)s")
		params["sales_person_user"] = filters["sales_person_user"]

	where = " AND ".join(conds)

	rows = frappe.db.sql(
		f"""
		SELECT
			si.custom_sales_person_user                                       AS sales_person,
			COUNT(DISTINCT si.name)                                           AS invoice_count,
			ROUND(SUM(si.grand_total), 2)                                     AS invoiced_amount,
			ROUND(SUM(si.grand_total - si.outstanding_amount), 2)             AS collected_amount,
			ROUND(SUM(si.outstanding_amount), 2)                              AS outstanding,
			CASE WHEN SUM(si.grand_total) > 0
			     THEN ROUND(SUM(si.grand_total - si.outstanding_amount)
			                / SUM(si.grand_total) * 100, 1)
			     ELSE 0
			END                                                               AS collection_pct
		FROM `tabSales Invoice` si
		WHERE {where}
		GROUP BY si.custom_sales_person_user
		ORDER BY collected_amount DESC
		""",
		params,
		as_dict=True,
	)
	return rows


def _chart(data, filters=None):
	chart_type = (filters or {}).get("chart_type", "bar")
	labels     = [r.sales_person for r in data]
	return {
		"type":    chart_type,
		"data": {
			"labels":   labels,
			"datasets": [
				{"name": _("Invoiced"),    "values": [flt(r.invoiced_amount)  for r in data]},
				{"name": _("Collected"),   "values": [flt(r.collected_amount) for r in data]},
				{"name": _("Outstanding"), "values": [flt(r.outstanding)      for r in data]},
			],
		},
		"colors": ["#5e64ff", "#28a745", "#e74c3c"],
	}


def _summary(data):
	total_inv   = sum(flt(r.invoiced_amount)  for r in data)
	total_col   = sum(flt(r.collected_amount) for r in data)
	total_out   = sum(flt(r.outstanding)      for r in data)
	overall_pct = round(total_col / total_inv * 100, 1) if total_inv else 0

	return [
		{"label": _("Total Invoiced"),    "value": total_inv,   "datatype": "Currency", "currency": "INR"},
		{"label": _("Total Collected"),   "value": total_col,   "datatype": "Currency", "currency": "INR"},
		{"label": _("Total Outstanding"), "value": total_out,   "datatype": "Currency", "currency": "INR"},
		{"label": _("Overall Collection %"), "value": overall_pct, "datatype": "Percent"},
	]

File: instabiz/instabiz/report/ib_collections_report/ib_collections_report.js
frappe.query_reports["IB Collections Report"] = {
	filters: [
		{
			fieldname: "from_date",
			label:     __("From Date"),
			fieldtype: "Date",
			default:   frappe.datetime.month_start(),
			reqd:      1,
		},
		{
			fieldname: "to_date",
			label:     __("To Date"),
			fieldtype: "Date",
			default:   frappe.datetime.get_today(),
			reqd:      1,
		},
		{
			fieldname: "territory",
			label:     __("State"),
			fieldtype: "Link",
			options:   "Territory",
		},
		{
			fieldname: "sales_person_user",
			label:     __("Sales Person"),
			fieldtype: "Link",
			options:   "User",
		},
		{
			fieldname: "chart_type",
			label:     __("Chart Type"),
			fieldtype: "Select",
			options:   "bar\npie\ndonut\nline\npercentage",
			default:   "bar",
		},
	],

	after_render() {
		const legend = this.$report?.find?.(".chart-legend");
		if (legend) legend.css("flex-wrap", "wrap");
	},

	formatter(value, row, column, data) {
		if (column.fieldname === "collection_pct" && data) {
			const pct = parseFloat(value) || 0;
			const color = pct >= 75 ? "green" : pct >= 40 ? "orange" : "red";
			return `<span style="color:var(--${color})">${flt(pct, 1)}%</span>`;
		}
		return value;
	},
};
